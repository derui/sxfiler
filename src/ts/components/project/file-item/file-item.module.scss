@import "../../ui/commom-style/_colors.scss";
@import "../../ui/commom-style/_variables.scss";
@import "../../ui/commom-style/reset.module.css";

.fileItem {
  composes: listItem-base from "../../ui/list-item/list-item.module.scss";

  padding: $baseSpace 0;
  width: 100%;

  background-color: $sol_base03;

  font-family: monospace;
  color: $sol_base1;
  white-space: nowrap;

  font-size: 1rem;
  border-left: 4px solid transparent;

  &[aria-selected="true"] {
    border-left: 4px solid $sol_orange;
  }

  &[aria-hidder="true"] {
    visibility: hidden;
  }

  &[data-marked="true"] {
    background-color: rgba($sol_blue, 0.25);
  }
}


.timestamp {
  flex: 0 1 auto;
  padding: 0 $smallSpace;
  text-align: right;

  white-space: nowrap;
}

.name {
  flex: 1 0 auto;
  padding: 0;
  font-weight: bold;
}

.name.nameDirectory {
  color: $sol_yellow;
}

.name.nameSymlink {
  color: $sol_orange;
}

.size {
  flex: 0 1 auto;
  padding: 0 $baseSpace;
  text-align: right;

  white-space: pre;
}

.mode {
  flex: 0 1 auto;
  padding: 0 $smallSpace;

  white-space: nowrap;
}
