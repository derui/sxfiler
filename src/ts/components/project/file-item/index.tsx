import { h } from "preact";

import { Component as Mode } from "./mode-slot";
import { Component as Name } from "./name-slot";
import { Component as Size } from "./size-slot";
import { Component as Timestamp } from "./timestamp-slot";
import { FileItem } from "@/generated/filer_pb";

export type Props = {
  item: FileItem.AsObject;
  selected: boolean;
  hidden?: boolean;
  bookmarked: boolean;
  onRefUpdated?: (e: HTMLElement) => void;
};

export const Component: preact.FunctionComponent<Props> = ({
  item,
  selected,
  bookmarked,
  hidden = false,
  onRefUpdated = undefined,
  ...rest
}) => {
  const stat = item.stat;
  if (!stat) {
    return null;
  }

  return (
    <div
      class="file-item__root"
      data-testid="fileList-item"
      aria-selected={selected}
      data-marked={item.marked}
      data-bookmarked={bookmarked}
      aria-hidden={hidden}
      ref={(e) => {
        if (onRefUpdated && e) {
          onRefUpdated(e);
        }
      }}
      {...rest}
    >
      <Mode key="mode" mode={stat.mode} isDirectory={stat.isDirectory} isSymlink={stat.isSymlink} />
      <Timestamp key="timestamp" timestamp={new Date(stat.mtime)} />
      <Size key="size" size={BigInt(stat.size)} />
      <Name key="name" name={item.name} isDirectory={stat.isDirectory} isSymlink={stat.isSymlink} />
    </div>
  );
};
Component.displayName = "FileItem";

export type ComponentType = typeof Component;
