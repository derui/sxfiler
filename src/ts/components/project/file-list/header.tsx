import * as React from "react";
import { styled } from "@/components/theme";

export type HeaderProps = {
  directory: string;
  focused: boolean;
};

const HeaderElement = styled.header`
  flex: 0 0 auto;
  padding: ${props => props.theme.spaces.base} 0;
  margin: 0;

  font-weight: bold;
  color: ${props => props.theme.colors.base1};
  background-color: ${props => props.theme.colors.base03};
  font-family: monospace;

  border-bottom: 1px solid ${props => props.theme.colors.base01};

  &[data-focused="true"] {
    color: ${props => props.theme.colors.yellow};
  }
`;

/**
 * component definition for header of file list
 */
export const Component: React.FunctionComponent<HeaderProps> = ({ directory, focused }) => {
  return <HeaderElement data-focused={focused}>{directory}</HeaderElement>;
};
