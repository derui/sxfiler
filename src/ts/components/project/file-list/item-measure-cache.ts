export type Measure = {
  width: number;
  height: number;
};

type Cache = Map<number, Measure>;
type ReadonlyCache = ReadonlyMap<number, Measure>;
type Handler = (cache: ReadonlyCache) => void;

export class ItemMeasureCache {
  private cache: Cache = new Map();
  private handlers: Handler[] = [];

  // regist handler to observe cache changes.
  public observe(handler: Handler): () => void {
    this.handlers.push(handler);
    return () => {
      this.handlers = this.handlers.filter((v) => v !== handler);
    };
  }

  public set(key: number, e: HTMLElement | null) {
    if (!e) {
      return;
    }

    const measure = {
      width: e.clientWidth,
      height: e.clientHeight,
    };

    this.cache.set(key, measure);

    const updatedCache = new Map(this.cache);
    this.handlers.forEach((v) => v(updatedCache));
  }

  public get(key: number) {
    return this.cache.get(key);
  }

  public clear() {
    this.cache.clear();
  }
}
