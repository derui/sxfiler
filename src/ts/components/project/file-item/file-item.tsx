import * as React from "react";
import { styled } from "@/components/theme";

import * as Domain from "@/domains/file-item";
import * as ListItem from "@/components/ui/list-item";
import { Component as Mode } from "./mode-slot";
import { Component as Name } from "./name-slot";
import { Component as Size } from "./size-slot";
import { Component as Timestamp } from "./timestamp-slot";
import { sizeAsBigInt } from "@/domains/file-stat";

export type Props = ListItem.Props & {
  item: Domain.FileItem;
  selected: boolean;
  hidden?: boolean;
  bookmarked: boolean;
};

const Element = styled(ListItem.Component)`
${ListItem.style}
  padding: ${props => props.theme.spaces.small} 0;
  width: 100%;

  background-color: ${props => props.theme.colors.base03};

  font-family: monospace;
  color: ${props => props.theme.colors.base2};
  white-space: nowrap;

  font-size: 1rem;
  border-left: 4px solid transparent;

  &[aria-selected="true"] {
    border-left: 4px solid ${props => props.theme.colors.orange};
  }

  &[aria-hidden="true"] {
    visibility: hidden;
  }

  &[data-marked="true"] {
    background-color: ${props => props.theme.colors.blue}3d;
  }

&[data-bookmarked="true"]{
    background-color: ${props => props.theme.colors.cyan}3d;
}
`;

const render = function renderFileItem(
  { item, selected, bookmarked, hidden = false, ...rest }: Props,
  ref: React.Ref<HTMLElement>
) {
  return (
    <Element
      selected={selected}
      data-marked={item.marked}
      data-bookmarked={bookmarked}
      aria-hidden={hidden}
      ref={ref}
      {...rest}
    >
      <Mode key="mode" mode={item.stat.mode} isDirectory={item.stat.isDirectory} isSymlink={item.stat.isSymlink} />
      <Timestamp key="timestamp" timestamp={item.stat.mtime} />
      <Size key="size" size={sizeAsBigInt(item.stat)} />
      <Name key="name" name={item.name} isDirectory={item.stat.isDirectory} isSymlink={item.stat.isSymlink} />
    </Element>
  );
};
render.displayName = "FileItem";

export const Component = React.forwardRef(render);

export type ComponentType = React.ComponentType<Props & React.RefAttributes<HTMLElement>>;
