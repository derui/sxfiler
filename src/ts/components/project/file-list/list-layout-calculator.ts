// Implement simple layouter to use render items of list to fill list based on curser on the list
import { ItemMeasureCache } from "./item-measure-cache";

/**
   A type saved item indiecs to show in a list. Each indices are zero-origin, and the range shows as `[start, stop)`
 */
export type VirtualizedWindow = {
  startIndex: number;
  stopIndex: number;
};

type InstanceProps = {
  estimatedItemSize: number;
};

type CalculationArguments = {
  cache: ItemMeasureCache;
  /**
   * The index of the cursor in the list. This value must be zero-origin
   */
  currentCursorIndex: number;
  windowHeight: number;
  listSize: number;
};

export class ListLayoutCalculator {
  private previousWindow: VirtualizedWindow | undefined;
  private instanceProps: InstanceProps = {
    estimatedItemSize: 50,
  };

  constructor(args: { estimatedItemSize?: number } = {}) {
    if (args.estimatedItemSize) {
      this.instanceProps.estimatedItemSize = args.estimatedItemSize;
    }
  }

  // get index range to show full of the body of the item in current window
  private resolveStartIndexForVisibleWindow(
    cache: ItemMeasureCache,
    listSize: number,
    windowHeight: number,
    currentCursorIndex: number
  ): number {
    let { startIndex } = this.previousWindow ? this.previousWindow : { startIndex: 0 };

    if (startIndex > currentCursorIndex) {
      return currentCursorIndex;
    } else {
      while (startIndex < currentCursorIndex) {
        let index;
        let restSize = windowHeight;
        for (index = startIndex; index < listSize; index++) {
          const measure = cache.get(index);
          let itemHeight = this.instanceProps.estimatedItemSize;

          if (measure) {
            itemHeight = measure.height;
          }
          if (restSize < itemHeight) {
            break;
          }

          restSize -= itemHeight;
        }

        if (startIndex <= currentCursorIndex && currentCursorIndex < index) {
          break;
        }

        startIndex = Math.min(startIndex + 1, listSize - 1);
      }
      return startIndex;
    }
  }

  private resolveStopIndexForStartIndex(
    cache: ItemMeasureCache,
    windowHeight: number,
    startIndex: number,
    listSize: number
  ): number {
    let restSize = windowHeight;
    for (let index = startIndex; index < listSize; index++) {
      if (restSize <= 0) {
        return index;
      }

      const measure = cache.get(index);
      let itemHeight = this.instanceProps.estimatedItemSize;

      if (measure) {
        itemHeight = measure.height;
      }
      restSize -= itemHeight;
    }

    return listSize;
  }

  // calculate layout of the list
  calculateLayout({ cache, currentCursorIndex, windowHeight, listSize }: CalculationArguments): VirtualizedWindow {
    const startIndex = this.resolveStartIndexForVisibleWindow(cache, listSize, windowHeight, currentCursorIndex);
    const stopIndex = this.resolveStopIndexForStartIndex(cache, windowHeight, startIndex, listSize);

    this.previousWindow = {
      startIndex,
      stopIndex,
    };
    return this.previousWindow;
  }

  // reset cached
  reset() {
    this.previousWindow = undefined;
  }
}
