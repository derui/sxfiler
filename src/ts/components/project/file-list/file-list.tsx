import * as React from "react";
import shallowequal from "shallowequal";
import { styled } from "@/components/theme";

import { FileItem } from "@/domains/file-item";
import * as Element from "@/components/ui/element";
import * as List from "@/components/ui/list";
import * as ListItem from "@/components/project/file-item";
import { AutoSizer } from "@/libs/auto-sizer";

import { ItemMeasureCache } from "./item-measure-cache";
import { ListLayoutCalculator, VirtualizedWindow } from "./list-layout-calculator";
import { Component as Header } from "./header";

export type Props = {
  location: string;
  items: FileItem[];
  cursor: number;
  focused: boolean;
};

const Root = styled(Element.Component)`
  display: grid;
  grid-template-rows: calc(${props => props.theme.spaces.base} * 2 + 1rem) 1fr;

  margin: 0;
  padding: 0;

  color: ${props => props.theme.colors.base01};
  overflow: hidden;
  background-color: ${props => props.theme.colors.base03};

  font-size: 1rem;
`;

const AutoSizerWrapper = styled(AutoSizer)`
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
`;

const Empty = styled.div`
  flex: 1 1 auto;
  margin: auto 0;
  text-align: center;

  color: ${props => props.theme.colors.red};

  &:after {
    content: "no any node";
    font-family: monospace;
  }
`;

const ListRoot = styled(List.Component)`
  ${List.style}
  position: relative;

  font-size: 1rem;
  border-right: 1px solid ${props => props.theme.colors.base0};

  &:last-of-type {
    border-right: none;
  }
`;

export type ElementType = React.ReactElement<Props, React.FunctionComponent<Props>>;

export class Component extends React.Component<Props> {
  private itemMeasureCache = new ItemMeasureCache();
  private layoutCalculator: ListLayoutCalculator = new ListLayoutCalculator({
    estimatedItemSize: 24,
  });

  constructor(props: Props) {
    super(props);
  }

  public shouldComponentUpdate(newProps: Props): boolean {
    if (!shallowequal(newProps, this.props)) {
      return true;
    }

    return false;
  }

  private makeList(height: number): JSX.Element {
    const { cursor } = this.props;
    const layout = this.layoutCalculator.calculateLayout({
      cache: this.itemMeasureCache,
      currentCursorIndex: cursor,
      windowHeight: height,
      listSize: this.props.items.length,
    });

    return (
      <ListRoot style={{ height: height }} key="body">
        {this.makeListItems(layout)}
      </ListRoot>
    );
  }

  private makeListItems(layout: VirtualizedWindow): JSX.Element[] {
    const { items, cursor, focused } = this.props;

    return items.slice(layout.startIndex, layout.stopIndex).map((item, index) => {
      const selected = cursor === index + layout.startIndex && focused;
      return (
        <ListItem.Component
          key={index + layout.startIndex}
          ref={e => this.itemMeasureCache.set(index + layout.startIndex, e)}
          item={item}
          selected={selected}
        />
      );
    });
  }

  public render() {
    const { items, focused } = this.props;

    return (
      <Root>
        <Header key="header" directory={this.props.location} focused={focused} />
        <AutoSizerWrapper key="sizer">
          {({ height }) => {
            return items.length === 0 ? <Empty /> : this.makeList(height);
          }}
        </AutoSizerWrapper>
      </Root>
    );
  }
}
