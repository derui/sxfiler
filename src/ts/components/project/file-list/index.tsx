import { h } from "preact";

import * as ListItem from "@/components/project/file-item";
import { AutoSizer } from "@/libs/auto-sizer";

import { ItemMeasureCache } from "./item-measure-cache";
import { ListLayoutCalculator, VirtualizedWindow } from "./list-layout-calculator";
import { FileItem } from "@/generated/filer_pb";
import { Bookmark } from "@/generated/bookmark_pb";
import { useState } from "preact/hooks";

export type Props = {
  location: string;
  items: FileItem.AsObject[];
  bookmarks: Bookmark.AsObject[];
  cursor: number;
  focused: boolean;
};

function useItemMeasureCache() {
  const [cache] = useState(new ItemMeasureCache());
  return cache;
}

const useLayoutCalculator = () => {
  const [calculator] = useState(
    new ListLayoutCalculator({
      estimatedItemSize: 24,
    })
  );

  return calculator;
};

const makeList = (
  height: number,
  props: Props,
  layoutCalculator: ListLayoutCalculator,
  itemMeasureCache: ItemMeasureCache
) => {
  const layout = layoutCalculator.calculateLayout({
    cache: itemMeasureCache,
    currentCursorIndex: props.cursor,
    windowHeight: height,
    listSize: props.items.length,
  });

  return (
    <div class="file-list__list-root" data-testid="fileList-list" style={{ height: height }} key="body">
      {makeListItems(layout, props, itemMeasureCache)}
    </div>
  );
};

const makeListItems = (layout: VirtualizedWindow, props: Props, itemMeasureCache: ItemMeasureCache) => {
  let keyAsPathBookmark = props.bookmarks.reduce((accum, v) => {
    accum[v.path] = v;
    return accum;
  }, {} as { [key: string]: Bookmark.AsObject });

  return props.items.slice(layout.startIndex, layout.stopIndex).map((item, index) => {
    const selected = props.cursor === index + layout.startIndex && props.focused;

    return (
      <ListItem.Component
        key={index + layout.startIndex}
        onRefUpdated={(e: HTMLElement) => itemMeasureCache.set(index + layout.startIndex, e)}
        item={item}
        bookmarked={!!keyAsPathBookmark[item.fullPath]}
        selected={selected}
      />
    );
  });
};

export type HeaderProps = {
  directory: string;
  focused: boolean;
};

/**
 * component definition for header of file list
 */
const Header: preact.FunctionComponent<HeaderProps> = ({ directory, focused }) => {
  return (
    <header class="file-list__header" data-testid="fileList-header" data-focused={focused}>
      {directory}
    </header>
  );
};

export const Component: preact.FunctionComponent<Props> = (props) => {
  const { items, focused, location } = props;
  const cache = useItemMeasureCache();
  const calculator = useLayoutCalculator();

  return (
    <div class="file-list__root" data-testid="fileList-root">
      <Header key="header" directory={location} focused={focused} />
      <AutoSizer class="file-list__autosizer" key="sizer">
        {({ height }) => {
          return items.length === 0 ? (
            <div class="file-list__empty" data-testid="fileList-empty" />
          ) : (
            makeList(height, props, calculator, cache)
          );
        }}
      </AutoSizer>
    </div>
  );
};

export type ElementType = ReturnType<typeof Component>;
