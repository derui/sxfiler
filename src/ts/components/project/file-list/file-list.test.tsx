import * as React from "react";
import renderer from "react-test-renderer";

import { Component as T } from "./file-list";

import { createFileItem } from "@/domains/file-item";
import { createFileStat } from "@/domains/file-stat";
import { createMode } from "@/domains/mode";
import { emptyCapability } from "@/domains/capability";

function makeNode(name: string, isDirectory = false, isSymlink = false) {
  return createFileItem({
    id: "node",
    name,
    marked: false,
    stat: createFileStat({
      mode: createMode({
        owner: emptyCapability()
          .allowToRead()
          .allowToWrite(),
        group: emptyCapability().allowToRead(),
        others: emptyCapability().allowToRead(),
      }),
      uid: 1000,
      gid: 1000,
      atime: "0",
      ctime: "0",
      mtime: "0",
      size: "10",
      isDirectory: isDirectory,
      isFile: !isDirectory && !isSymlink,
      isSymlink: isSymlink,
    }),
    parentDirectory: "/",
  });
}

describe("Project", () => {
  describe("Node List", () => {
    it("should not print before resized", () => {
      const nodes = [makeNode("file.txt")];
      const tree = renderer.create(<T items={nodes} location="loc" cursor={0} focused={false} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should select a node locating same the index of cursor when focused", () => {
      const nodes = [makeNode("file.txt")];
      const tree = renderer.create(<T items={nodes} location="loc" cursor={0} focused={true} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should show dummy content when nodes is empty", () => {
      const tree = renderer.create(<T items={[]} location="loc" cursor={0} focused={true} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
