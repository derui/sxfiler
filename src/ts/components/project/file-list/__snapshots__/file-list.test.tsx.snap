// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Node List should not print before resized 1`] = `
<div
  className="sc-bZQynM gAHUms"
>
  <header
    className="sc-EHOje eksHn"
    data-focused={false}
  >
    loc
  </header>
  <div
    className="sc-gzVnrw VQBHp"
    style={
      Object {
        "overflow": "visible",
      }
    }
  />
</div>
`;

exports[`Project Node List should select a node locating same the index of cursor when focused 1`] = `
<div
  className="sc-bZQynM gAHUms"
>
  <header
    className="sc-EHOje eksHn"
    data-focused={true}
  >
    loc
  </header>
  <div
    className="sc-gzVnrw VQBHp"
    style={
      Object {
        "overflow": "visible",
      }
    }
  />
</div>
`;

exports[`Project Node List should show dummy content when nodes is empty 1`] = `
<div
  className="sc-bZQynM gAHUms"
>
  <header
    className="sc-EHOje eksHn"
    data-focused={true}
  >
    loc
  </header>
  <div
    className="sc-gzVnrw VQBHp"
    style={
      Object {
        "overflow": "visible",
      }
    }
  />
</div>
`;
